### DISCUSSION TITLE
More work on MD buttons

### DISCUSSION THREAD

**Discussion by gbm001** [HeaTTheatR/KivyMD#1145] [General]
This is to keep track, and allow comment on, things I think either need fixing or I am planning on improving to do with KivyMD Buttons.

- If you use a multi-line label (which of course is not allowed by the MD spec, but that doesn't mean you can't) the label is not centered (easy fix - set halign='center', valign='center'/'middle' or whatever it is on the label).
- There may be some issues with square backgrounds behind FABs on Android (but not Linux) - I need to check this (might just be weirdness with my own app).
- Button sizes are not in accordance with the MD spec. The padding attribute is currently ignored (if symmetric) rather than added to the height/width. If the padding attribute is not symmetric then weirdness happens. The minimum height for buttons is currently set at 88 but the MD2 spec is 64dp (48dp in MD3). The button height is fixed in MD2 at 36dp (40dp in MD3); we should allow buttons to become larger than this when required but for text that fits this should be the button height. For all except text-only buttons, the padding should be 12dp to the icon (whether present or not), 4dp to the text label, and 16dp on the far side (16dp, 8dp and 24dp, respectively, for MD3). Text buttons should only have 8dp either side (12dp in MD3, with 8dp between icon and text if both present).
- Right-to-left text should have the icon at the right-hand end of an icon & text button.
- MD3 buttons are not supported (except FABs).

There are fundamentally two kinds of (MD2) buttons currently supported:
- 'Normal' buttons: These can be simple text buttons, or can be outlined, or can be filled, can have elevation behaviour, and can be rounded rectangles or just rectangles.
- 'Icon' buttons: Circular buttons with an icon at the centre; can be a FloatingActionButton if it has elevation behaviour.

In both MD2 and MD3, Floating Action Buttons can have text as well as an icon ('extended' FABs). This leaves the plain icon button as the _only_ button that doesn't take potentially both text and an icon. Consequently, it would probably be easier to roll the Icon properties into BaseButton since they could be used in almost every MD button, and always include (at least for the 'rectangular' buttons) both an icon and a text label. If the button had no icon (for a 'rectangular' button), the label would just be zero-width which (except for text-only buttons) gets the correct padding.
Consequently, I propose making these changes, fixing the label size/padding issues and minimum width/heights for the existing classes, then creating an 'MDButton' class which can be set to MD2 or MD3 and covers all rectangular button options via a 'button_type' parameter (which sets defaults for line color/width, background color/width etc). Then MDIconButton and MDFloatingActionButton would then remain as separate classes.

**Comment by HeaTTheatR** [comment]
@gbm001 In my opinion, this is a more appropriate place for this topic :)

**Reply by gbm001** [reply]
You learn something new every day - I didn't even realise GitHub did discussions :)

**Comment by HeaTTheatR** [comment]
@gbm001 

```python
from kivy.lang import Builder

from kivymd.app import MDApp

KV = '''

MDScreen:

    MDIconButton:
        size_hint: None, None
        pos_hint: {"center_y": 0.5}
        size: 0, 0
'''


class Test(MDApp):
    def build(self):
        return Builder.load_string(KV)



Test().run()
```

```python
ValueError: MDIconButton._radius is below the minimum bound (0.100000)
```

https://github.com/kivymd/KivyMD/blob/master/kivymd/uix/button/button.py#L861

**Comment by HeaTTheatR** [comment]
For example, similar errors occur when using the `MDDataTable` class.

**Reply by kevinlaurencedeffo** [reply]
about this class I have a problem

**Reply by kevinlaurencedeffo** [reply]
File "kivy\graphics\vertex_instructions.pyx", line 1574, in kivy.graphics.vertex_instructions.RoundedRectangle.radius.__set__
   File "kivy\graphics\vertex_instructions.pyx", line 1334, in kivy.graphics.vertex_instructions.RoundedRectangle._check_radius
 TypeError: 'NoneType' object is not iterable

 During handling of the above exception, another exception occurred:

 Traceback (most recent call last):
   File "C:\Users\king\Documents\doc\python\gestion\main.py", line 164, in <module>
     Main().run()
   File "C:\Program Files\Python39\lib\site-packages\kivy\app.py", line 954, in run
     self._run_prepare()
   File "C:\Program Files\Python39\lib\site-packages\kivy\app.py", line 924, in _run_prepare
     root = self.build()
   File "C:\Users\king\Documents\doc\python\gestion\main.py", line 121, in build
     data = AnchorLayout(MDDataTable(data_table_bd = MDDataTable(
   File "C:\Program Files\Python39\lib\site-packages\kivymd\uix\datatables.py", line 1555, in __init__
     self.header = TableHeader(
   File "C:\Program Files\Python39\lib\site-packages\kivymd\uix\datatables.py", line 635, in __init__
     super().__init__(**kwargs)
   File "C:\Program Files\Python39\lib\site-packages\kivy\uix\scrollview.py", line 539, in __init__
     super(ScrollView, self).__init__(**kwargs)
   File "C:\Program Files\Python39\lib\site-packages\kivy\uix\widget.py", line 366, in __init__
     self.apply_class_lang_rules(
   File "C:\Program Files\Python39\lib\site-packages\kivy\uix\widget.py", line 470, in apply_class_lang_rules        
     Builder.apply(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 540, in apply
     self._apply_rule(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 662, in _apply_rule
     self._apply_rule(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 662, in _apply_rule
     self._apply_rule(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 662, in _apply_rule
     self._apply_rule(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 660, in _apply_rule
     child.apply_class_lang_rules(
   File "C:\Program Files\Python39\lib\site-packages\kivy\uix\widget.py", line 470, in apply_class_lang_rules        
     Builder.apply(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 540, in apply
     self._apply_rule(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 660, in _apply_rule
     child.apply_class_lang_rules(
   File "C:\Program Files\Python39\lib\site-packages\kivy\uix\widget.py", line 470, in apply_class_lang_rules        
     Builder.apply(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 540, in apply
     self._apply_rule(
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 595, in _apply_rule
     self._build_canvas(widget.canvas.before, widget,
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 928, in _build_canvas
      15:            size: self.size
      16:            pos: self.pos if not isinstance(self, RelativeLayout) else (0, 0)
 >>   17:            radius: root.radius
      18:            source: root.background
      19:        Color:
 ...
 TypeError: 'NoneType' object is not iterable
   File "C:\Program Files\Python39\lib\site-packages\kivy\lang\builder.py", line 925, in _build_canvas
     setattr(instr, key, value)
   File "kivy\graphics\vertex_instructions.pyx", line 1574, in kivy.graphics.vertex_instructions.RoundedRectangle.radius.__set__
   File "kivy\graphics\vertex_instructions.pyx", line 1334, in kivy.graphics.vertex_instructions.RoundedRectangle._check_radius

**Reply by gbm001** [reply]
This (the middle of a discussion about buttons) probably isn't a place to submit a bug report about datatables?
More information would be needed in any event.

**Comment by gbm001** [comment]
I've obviously done the error handler on BoundedNumericProperty for _radius wrong; I wanted it to automatically set anything less than 0.1 to 0.1.

**Reply by gbm001** [reply]
I can fix this tomorrow...

**Reply by gbm001** [reply]
Fixed by #1150 . I reckon this isn't really my fault; it's horrible floating point behaviour from Kivy... (you should be able to set an error value that is equal to one of your supposedly inclusive bounds, even if that value is not perfectly representable in floating point).

**Comment by HeaTTheatR** [comment]
@gbm001 Perhaps this bug should be reported in Kivy issues.

**Comment by gbm001** [comment]
Is there a plan to implement an MD3 colour scheme into KivyMD? In theory, it should be simpler even than using the MD2014 colour themes; you just pick a primary, secondary and tertiary hue, a neutral and neutral variant hue, and an error hue, and then just magically generate everything from that (I say 'hue' because I think the primary/secondary/tertiary etc. should be 40% (?) shades).

**Reply by HeaTTheatR** [reply]
I have not yet been able to understand the benefits of the MD3 color themes.

**Reply by gbm001** [reply]
> I have not yet been able to understand the benefits of the MD3 color themes.

'Tonal' buttons in MD3 are defined as being on the 'Secondary Container' colour which is a paler version of the secondary colour ('accent' in KivyMD?) which doesn't really have an equivalent in MD2 I think, meaning you either have to bodge the MD3 buttons with MD2 colour variants that are similar or implement the MD3 colours.

**Comment by HeaTTheatR** [comment]
@gbm001 

Added the ability to set a custom disabled button color for icon elements, text and button line:

![Снимок экрана 2021-12-24 в 15 35 14](https://user-images.githubusercontent.com/16930280/147353038-505a4080-c5c9-4929-bb65-c0c1050cd8ba.png)


```python
from kivy.lang import Builder

from kivymd.app import MDApp


KV = '''
MDScreen:
        
    MDRoundFlatIconButton:
        text: "Button disabled with default disabled color"
        disabled: True
        pos_hint: {"center_x": .5, "center_y": .2}
        
    MDRoundFlatIconButton:
        id: btn
        text: "Button disabled with custom disabled color"
        disabled: True
        disabled_color: 0, 0, 1, 1
        pos_hint: {"center_x": .5, "center_y": .3}
        
    MDRaisedButton:
        text: "Set disabled color"
        pos_hint: {"center_x": .5, "center_y": .4}
        on_release: btn.disabled_color = (1, 1, 0, 1)
'''


class Test(MDApp):
    def build(self):
        return Builder.load_string(KV)


Test().run()
```

https://github.com/kivymd/KivyMD/compare/9fdf5e226908...c1c9465f56c1

**Reply by gbm001** [reply]
Very reasonable... I am also starting to think that the 'correct' method of setting the disabled colour based on 38% of the 'valid' colour is silly (since it leads to differently-coloured buttons having different disabled colours).

Incidentally that _isn't_ a thing in the MD3 colour schemes, where the disabled colour is set relative to the 'surface' (background?) colour (or something like that).

I do think the MD3 colour scheme approach (that you have not only a primary colour, but also the colour of text/icons on that primary colour, and the colour of a primary container which is paler, and the colour of text/icons on that container colour) is probably easier to use? I may knock up a class to define all the relevant colours.

**Comment by HeaTTheatR** [comment]
@gbm001 
The MD3 color theme isn't just about buttons. This is a very huge job, which implies a fundamentally different approach to working with the palette. I don’t know how to implement it correctly yet.

**Reply by gbm001** [reply]
Agreed... I can (if I get a chance) knock up something to generate all the colours, but I have no idea how that would be best integrated into KivyMD in parallel with the existing colour system.

**Comment by JUNIORjr250** [comment]
Hello I'm New

**Reply by MudrykD5** [reply]
yooo

**Comment by MudrykD5** [comment]
hello everyone, i need help on adding a textfield in kivymd dialog in version 2.0.1.dev0
